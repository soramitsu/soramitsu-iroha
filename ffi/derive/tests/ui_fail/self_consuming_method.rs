//! Test which assures that methods consuming self are not allowed in FFI. If the need arises,
//! this limitation can be lifted in the future, and consequently this test nullified

use iroha_ffi::{ffi_bindgen, FfiResult};
use std::mem::MaybeUninit;

struct FfiStructBuilder;
struct FfiStruct;

#[ffi_bindgen]
impl FfiStructBuilder {
    /// New
    pub fn new() -> Self {
        Self
    }

    /// Build
    pub fn build(self) -> FfiStruct {
        FfiStruct
    }
}

fn main() -> Result<(), ()> {
    let s_builder: MaybeUninit<*mut FfiStructBuilder> = MaybeUninit::uninit();
    if FfiResult::Ok != FfiStructBuilder__new(s_builder.as_mut_ptr()) {
        return Err(());
    }

    let s_builder = unsafe { s_builder.assume_init() };
    let s: MaybeUninit<*mut FfiStruct> = MaybeUninit::uninit();
    if FfiResult::Ok != FfiStructBuilder__build(s_builder, s.as_mut_ptr()) {
        return Err(());
    }

    Ok(())
}
