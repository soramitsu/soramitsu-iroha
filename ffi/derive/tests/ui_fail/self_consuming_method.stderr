error: Methods which consume self not supported
  --> tests/ui_fail/self_consuming_method.rs:11:6
   |
11 | impl FfiStructBuilder {
   |      ^^^^^^^^^^^^^^^^

error[E0425]: cannot find function, tuple struct or tuple variant `FfiStructBuilder__new` in this scope
  --> tests/ui_fail/self_consuming_method.rs:25:25
   |
7  | struct FfiStructBuilder;
   | ------------------------ similarly named unit struct `FfiStructBuilder` defined here
...
25 |     if FfiResult::Ok != FfiStructBuilder__new(s_builder.as_mut_ptr()) {
   |                         ^^^^^^^^^^^^^^^^^^^^^ help: a unit struct with a similar name exists: `FfiStructBuilder`

error[E0425]: cannot find function, tuple struct or tuple variant `FfiStructBuilder__build` in this scope
  --> tests/ui_fail/self_consuming_method.rs:31:25
   |
7  | struct FfiStructBuilder;
   | ------------------------ similarly named unit struct `FfiStructBuilder` defined here
...
31 |     if FfiResult::Ok != FfiStructBuilder__build(s_builder, s.as_mut_ptr()) {
   |                         ^^^^^^^^^^^^^^^^^^^^^^^ help: a unit struct with a similar name exists: `FfiStructBuilder`
